### PR TITLE
add support for validating admission webhook in terrascan

### PR DESCRIPTION
**Pull request by kanchwala-yusuf**

### CODE REVIEW

File: pkg/http-server/start.go [hunk 1]
```diff
@@ -61,14 +62,24 @@ func (g *APIServer) start(routes []*Route, port string) {
 		router.Methods(v.verb).Path(v.path).HandlerFunc(v.fn)
 	}
 
+	// Add a route for all static templates / assets. Currently used for the Webhook logs views
+	// go/terrascan/asset is the path where the assets files are located inside the docker container
+	router.PathPrefix("/assets/").Handler(http.StripPrefix("/assets/", http.FileServer(http.Dir("/go/terrascan/assets"))))
```
Comment (dev-gaur) [inline pkg/http-server/start.go:67]: hardcoding `/go/terrascan/assets` can cause potential bugs. maybe we can introduce an env variable for this directory path and keep this value as default.

File: docs/getting-started/admission-controller-webhooks-usage.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+* Scan configuration changes policies when an object is being created or updated
+* Allow / reject the request in case a violation is detected
+
+
+## Installation Guide
+
+### Create an instance
+Your Terrascan instance has the following requirements for being able to scan K8s configurations.
+
+1. Be accessible via HTTPS. Make sure your cloud firewall is configured to allow this.
```
Comment (cesar-rodriguez) [inline docs/getting-started/admission-controller-webhooks-usage.md:17]: The bullet numbers need to be fixed

**Review by cesar-rodriguez — CHANGES_REQUESTED** [review]

File: pkg/http-server/webhook-scan_test.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+	"os"
+	"testing"
+
+	"github.com/accurics/terrascan/pkg/k8s/dblogs"
+	"github.com/gorilla/mux"
+	v1 "k8s.io/api/admission/v1"
+)
+
+func TestUWebhooks(t *testing.T) {
+	testFilePath := "./k8s_testdata/testconfig.json"
```
Comment (patilpankaj212) [inline pkg/http-server/webhook-scan_test.go:19]: we should form all paths using `filepath.Join()`

File: pkg/k8s/admission-webhook/validating-webhook.go [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+
+	var (
+		executor *runtime.Executor
+		err      error
+		result   runtime.Output
+	)
+
+	if flag.Lookup("test.v") != nil {
+		executor, err = runtime.NewExecutor("k8s", "v1", []string{"k8s"},
+			filePath, "", w.configFile, []string{"../policies/opa/rego/k8s"}, []string{}, []string{}, []string{}, "")
```
Comment (patilpankaj212) [inline pkg/k8s/admission-webhook/validating-webhook.go:182]: we should use `filepath.Join()` for the test policy path

File: pkg/k8s/admission-webhook/validating-webhook_test.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+func TestDecodeAdmissionReviewRequest(t *testing.T) {
+
+	table := []struct {
+		name        string
+		requestFile string
+		wantErr     bool
+	}{
+		{
+			name:        "empty review request",
+			requestFile: "testdata/empty.json",
```
Comment (patilpankaj212) [inline pkg/k8s/admission-webhook/validating-webhook_test.go:88]: we should use `filepath.Join()` for all paths

**Review by patilpankaj212 — CHANGES_REQUESTED** [review]

**Review by dev-gaur — CHANGES_REQUESTED** [review]

File: pkg/http-server/webhook-scan-logs.go [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+	// 1. !Allowed -> Rejected
+	// 2. Allowed -> if there are violations -> Allowed with Warnings. Otherwise -> Allowed
+	if !log.Allowed {
+		return "Rejected"
+	}
+
+	var violationStore results.ViolationStore
+	err := json.Unmarshal([]byte(log.ViolationsSummary), &violationStore)
+	if err != nil {
+		zap.S().Errorf("Failed to ..")
```
[the remark this answers is not in the document]
Comment (kanchwala-yusuf) [reply pkg/http-server/webhook-scan-logs.go]: Done

File: pkg/cli/server.go [hunk 7]
```diff
@@ -21,6 +21,17 @@ import (
 	"github.com/spf13/cobra"
 )
 
+var (
+	// Port at which API server will listen
+	Port string
```
[the remark this answers is not in the document]
Comment (kanchwala-yusuf) [reply pkg/cli/server.go]: Done

File: pkg/http-server/webhook-scan-logs.go [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+
+	// Return an HTML page including all the logs history
+	logger := dblogs.NewWebhookScanLogger()
+
+	// The templates are saved in the docker in this location
+	t, err := template.ParseFiles("/go/terrascan/index.html")
+	if err != nil {
+		errMsg := fmt.Sprintf("failed to parse index.html file; error: '%v'", err)
+		zap.S().Error(errMsg)
+		apiErrorResponse(w, errMsg, http.StatusInternalServerError)
```
[the remark this answers is not in the document]
Comment (kanchwala-yusuf) [reply pkg/http-server/webhook-scan-logs.go:75]: Done

File: pkg/http-server/webhook-scan.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+	// Send the correct response according to the result
+	g.sendResponseAdmissionReview(w, admissionResponse)
+}
+
+func (g *APIHandler) sendResponseAdmissionReview(w http.ResponseWriter, admissionResponse *v1.AdmissionReview) {
+	respBytes, err := json.Marshal(admissionResponse)
+	if err != nil {
+		msg := fmt.Sprintf("failed to serialize admission review response: %v", err)
+		zap.S().Error(msg)
+		apiErrorResponse(w, msg, http.StatusInternalServerError)
```
[the remark this answers is not in the document]
Comment (kanchwala-yusuf) [reply pkg/http-server/webhook-scan.go:91]: Done

File: pkg/http-server/webhook-scan_test.go (hunk 3 above)
Comment (kanchwala-yusuf) [reply pkg/http-server/webhook-scan_test.go]: Done

File: pkg/k8s/admission-webhook/validating-webhook.go (hunk 4 above)
Comment (kanchwala-yusuf) [reply pkg/k8s/admission-webhook/validating-webhook.go]: Done

File: pkg/k8s/admission-webhook/validating-webhook_test.go (hunk 5 above)
Comment (kanchwala-yusuf) [reply pkg/k8s/admission-webhook/validating-webhook_test.go]: Done

File: pkg/http-server/start.go (hunk 1 above)
Comment (kanchwala-yusuf) [reply pkg/http-server/start.go:67]: Introducing a new environment variable may not be the best way to go ahead. This path is not something that an end user would be configuring. May be we can document that this endpoint would work on docker containers

File: docs/getting-started/admission-controller-webhooks-usage.md (hunk 2 above)
Comment (kanchwala-yusuf) [reply docs/getting-started/admission-controller-webhooks-usage.md:17]: done

File: docs/getting-started/admission-controller-webhooks-usage.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+      "Network Ports Security"
+  ]
+  denied-severity = "high"
+  ```
+
+You can specify the following configurations:
+*  **scan-rules** - one or more rules to scan
+*  **skip-rules** - one or more rules to skip while scanning
+*  **severity** - the minimal level of severity of the policies to be scanned
+
```
[the remark this answers is not in the document]
Comment (kanchwala-yusuf) [reply docs/getting-started/admission-controller-webhooks-usage.md:57]: done

File: pkg/http-server/handler.go [hunk 11]
```diff
@@ -18,10 +18,13 @@ package httpserver
...
+	test       bool
+	configFile string
 }
 
 // NewAPIHandler returns a new APIHandler{}
-func NewAPIHandler() *APIHandler {
-	return &APIHandler{}
+func NewAPIHandler(configFile string) *APIHandler {
+	return &APIHandler{
+		configFile: configFile,
```
[the remark this answers is not in the document]
Comment (kanchwala-yusuf) [reply pkg/http-server/handler.go:28]: The config file changes are quite significant and can be handled as separate PR. adding those changes in this PR would just increase the bulk of this PR.

If you are okay, we can open a separate issue for this and address it as a part of it.

File: pkg/k8s/admission-webhook/validating-webhook.go [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+
+func (w ValidatingWebhook) scanK8sFile(filePath string) (runtime.Output, error) {
+
+	var (
+		executor *runtime.Executor
+		err      error
+		result   runtime.Output
+	)
+
+	if flag.Lookup("test.v") != nil {
```
Comment (jlk) [inline pkg/k8s/admission-webhook/validating-webhook.go:185]: Not looking for a change, but this feels strange as almost all the code uses Cobra...

**Review by jlk — APPROVED** [review]
Overall LGTM. Wouldn't mind changing that CSS file, though.

**Review by patilpankaj212 — APPROVED** [review]
LGTM!
We should add e2e tests for admission webhook later.

File: pkg/http-server/templates/show.html [hunk 13]
```diff
@@ -0,0 +1,40 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <title>K8s Admission Review Logs</title>
+  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css" integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">
```
[the remark this answers is not in the document]
Comment (kanchwala-yusuf) [reply pkg/http-server/templates/show.html]: done!

**Review by dev-gaur — CHANGES_REQUESTED** [review]
requested a small change in docs

File: docs/getting-started/admission-controller-webhooks-usage.md [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+*  **category** - the list of type of categories of the policies to be scanned
+
+
+* **k8s-deny-rules** - specify the rules that should cause a rejection of the admission request
+  *  **denied-categories** - one or more policy categories that are not allowed in the detected violations
+  *  **denied-severity** - the minimal level of severity that should cause a rejection
+
+In order to use a configuration file, add it as a command line argument:
+
+```<terrascan> -c /data/my_terrscan_config.toml```
```
[the remark this answers is not in the document]
Comment (kanchwala-yusuf) [reply docs/getting-started/admission-controller-webhooks-usage.md]: updated the docker command in the documentation

**Review by cesar-rodriguez — APPROVED** [review]

File: pkg/http-server/handler.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (dev-gaur) [reply pkg/http-server/handler.go:28]: Sure. let's tackle this refactor seperately.

I'll also create one for helm charts.

**Review by dev-gaur — APPROVED** [review]

**Review by patilpankaj212 — APPROVED** [review]

**Pull request merged**